include/pisa/query.hpp

```diff
@@ -0,0 +1,173 @@
+// Copyright 2024 PISA Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <type_traits>
+#include <vector>
+
+#include "type_alias.hpp"
+
+namespace pisa {
+
+/** Term ID along with its weight.
+ *
+ * Typically, a weight would be equal to the number of occurrences of the term in a query.
+ * Partial scores coming from this term will be multiplied by this weight.
+ */
+struct WeightedTerm {
+    TermId id;
+    Score weight;
+
+    /** Tuple conversion for structured bindings support.
+     *
+     * ```
+     * WeightedTerm wt{0, 1.0};
+     * auto [term_id, weight] = weighted_term;
+     * ```
+     */
+    [[nodiscard]] explicit constexpr operator std::pair<TermId, Score>() const noexcept;
+};
+
+[[nodiscard]] inline auto operator==(WeightedTerm const& lhs, WeightedTerm const& rhs) -> bool {
+    return std::tie(lhs.id, lhs.weight) == std::tie(rhs.id, rhs.weight);
+}
+
+namespace query {
+
+    /**
+     * Tells `Query` how to process the terms passed to the constructor.
+     *
+     * By default, duplicate terms will be removed, and the weight of each term will be equal to
+     * the number of occurrences of that term in the query. Furthermore, the order of the terms
+     * will be preserved (if there are duplicates, the term will be at the position of its first
+     * occurrence.
+     *
+     * This policy can be modified with the following options:
+     *  - `keep_duplicates`: duplicates will be preserved, each with weight 1.0;
```

Comment:
Should we note that this is *very bad* in terms of efficiency?

Reply:
Good idea

Reply:
Check out my edit.

Reply:
Great, thanks a lot!